src/time_series_metadata_store.jl

```diff
@@ -0,0 +1,998 @@
+const METADATA_TABLE_NAME = "time_series_metadata"
+const DB_FILENAME = "time_series_metadata.db"
+
+mutable struct TimeSeriesMetadataStore
+    db::SQLite.DB
+end
+
+"""
+Construct a new TimeSeriesMetadataStore with an in-memory database.
+"""
+function TimeSeriesMetadataStore()
+    store = TimeSeriesMetadataStore(SQLite.DB())
+    _create_metadata_table!(store)
+    _create_indexes!(store)
+    @debug "Initialized new time series metadata table" _group = LOG_GROUP_TIME_SERIES
+    return store
+end
+
+"""
+Load a TimeSeriesMetadataStore from a saved database into an in-memory database.
+"""
+function TimeSeriesMetadataStore(filename::AbstractString)
+    src = SQLite.DB(filename)
+    db = SQLite.DB()
+    backup(db, src)
+    store = TimeSeriesMetadataStore(db)
+    @debug "Loaded time series metadata from file" _group = LOG_GROUP_TIME_SERIES filename
+    return store
+end
+
+"""
+Load a TimeSeriesMetadataStore from an HDF5 file into an in-memory database.
+"""
+function from_h5_file(::Type{TimeSeriesMetadataStore}, src::AbstractString, directory)
+    data = HDF5.h5open(src, "r") do file
+        file[HDF5_TS_METADATA_ROOT_PATH][:]
+    end
+
+    filename, io = mktemp(isnothing(directory) ? tempdir() : directory)
+    write(io, data)
+    close(io)
+    return TimeSeriesMetadataStore(filename)
+end
+
+function _create_metadata_table!(store::TimeSeriesMetadataStore)
+    # TODO: SQLite createtable!() doesn't provide a way to create a primary key.
+    # https://github.com/JuliaDatabases/SQLite.jl/issues/286
+    # We can use that function if they ever add the feature.
+    schema = [
+        "id INTEGER PRIMARY KEY",
+        "time_series_uuid TEXT NOT NULL",
+        "time_series_type TEXT NOT NULL",
+        "time_series_category TEXT NOT NULL",
+        "initial_timestamp TEXT NOT NULL",
+        "resolution_ms INTEGER NOT NULL",
+        "horizon INTEGER",
+        "horizon_time_ms INTEGER",
+        "interval_ms INTEGER",
+        "window_count INTEGER",
+        "length INTEGER",
+        "name TEXT NOT NULL",
+        "owner_uuid TEXT NOT NULL",
+        "owner_type TEXT NOT NULL",
+        "owner_category TEXT NOT NULL",
+        "features TEXT NOT NULL",
+        # The metadata is included as a convenience for serialization/de-serialization,
+        # specifically for types: time_series_type and scaling_factor_mulitplier.
+        # There is a lot duplication of data.
+        "metadata JSON NOT NULL",
```

Comment:
I tested the length of this field for a time series with two features. 459 bytes. Here is likely a worst-case system: 100,000 components each with 10 time series arrays each with 2 features. 
```
100_000 * 10 * 459 / (1024*1024)
437.73651123046875
```
437 MB wasted is likely not a big deal, but I'll prototype the alternative implementation just to see.